thread/common/threadhelpers.cpp

```diff
@@ -483,4 +483,96 @@ omrthread_mcs_node_free(omrthread_t self, omrthread_mcs_node_t mcsNode)
+ * by omrthread_park and omrthread_park_spin.
+ * @param[in] thread the current omrthread_t
+ *
+ * @return 0 if no checked bit is set
+ * J9THREAD_INTERRUPTED if the thread was interrupted while parked<br>
+ * J9THREAD_PRIORITY_INTERRUPTED if the thread was priority interrupted or aborted while parked<br>
+ * J9THREAD_UNPARKED if the thread has been unparked. Note: this return value is only for internal
+ * use and should not be returned by omrthread_park
+ */
+intptr_t
+omrthread_park_check_flags(omrthread_t thread)
+{
+	intptr_t rc = 0;
+	if (OMR_ARE_ANY_BITS_SET(thread->flags, J9THREAD_FLAG_UNPARKED)) {
+		rc = J9THREAD_UNPARKED;
+	} else if (OMR_ARE_ANY_BITS_SET(thread->flags, J9THREAD_FLAG_INTERRUPTED)) {
+		rc = J9THREAD_INTERRUPTED;
+	} else if (OMR_ARE_ANY_BITS_SET(thread->flags, (J9THREAD_FLAG_PRIORITY_INTERRUPTED | J9THREAD_FLAG_ABORTED))) {
+		rc = J9THREAD_PRIORITY_INTERRUPTED;
+	}
+	return rc;
+}
+
+#if defined(OMR_THR_YIELD_ALG)
+/**
+ * Helper function to be called by omrthread_park to spin or sleep the current thread before waiting
+ * on the condition variable.
+ *
+ * @param[in] self the current omrthread_t
+ * @param[in] millis millis argument of the calling omrthread_park
+ * @param[in] nanos nanos argument of the calling omrthread_park
+ * @param[out] sleptDuration the elapsed time of sleep when sleep count is exhausted or the thread
+ * is unparked
+ *
+ * @return 0 if the thread hasn't been unparked or timed out after spinning or sleeping
+ * J9THREAD_INTERRUPTED if the thread was interrupted while parked<br>
+ * J9THREAD_PRIORITY_INTERRUPTED if the thread was priority interrupted or aborted while parked<br>
+ * J9THREAD_TIMED_OUT if the timeout expired<br>
+ * J9THREAD_UNPARKED if the thread has been unparked. Note: this return value is only for internal
+ * use and should not be returned by omrthread_park
+ */
+intptr_t
+omrthread_park_spin(omrthread_t self, int64_t millis, intptr_t nanos, uintptr_t *sleptDuration)
+{
+	omrthread_library_t lib = self->library;
+	intptr_t rc = 0;
+	uintptr_t parkPolicy = lib->parkPolicy;
+	useconds_t totalSleepTime = 0;
+
+	if ((OMRTHREAD_PARK_POLICY_NONE != parkPolicy)) {
+		uintptr_t count = 0;
+		uintptr_t parkSleepTime = lib->parkSleepTime;
+		uintptr_t parkSleepMultiplier = lib->parkSleepMultiplier;
+		uintptr_t timeout = (millis * 1000) + (nanos / 1000);
+
+		if (OMRTHREAD_PARK_POLICY_SPIN == parkPolicy) {
+			/* For now, spin is only supported if no timeout is specified. */
+			if ((0 == millis) && (0 == nanos)) {
+				count = lib->parkSpinCount;
+			}
+		} else if (OMRTHREAD_PARK_POLICY_SLEEP == parkPolicy) {
+			count = lib->parkSleepCount;
+		}
+
+		for (uintptr_t i = 0; i < count; i++) {
+			rc = omrthread_park_check_flags(self);
+			if (0 != rc) {
+				break;
+			}
+			if (OMRTHREAD_PARK_POLICY_SLEEP == parkPolicy) {
+				useconds_t sleepTime = OMR_MIN(((i * parkSleepMultiplier) + 1) * parkSleepTime, 999999);
+				totalSleepTime += sleepTime;
+				usleep(sleepTime);
```

Comment:
> `int64_t` should be used for this calculation and the `usleep` calculation to avoid truncation.

All points have been addressed, except for the issue noted above. The same problem will also occur in the `usleep` calculation, since `totalSleepTime` is of type `useconds_t`.